include/flashinfer/comm/trtllm_mnnvl_allreduce.cuh

```diff
@@ -95,581 +96,1126 @@ inline __device__ __nv_half fromFloat<__nv_half>(float val) {
-
-        // Calculate the absolute element offset from the start of buffer_input
-        int element_offset = offsets[i][j];
-
-        // The input pointer is already offset to: &buffer_input[buffer_offset + buffer_size]
-        // So the actual pointer we're accessing is: input + element_offset
-        // Which equals: &buffer_input[buffer_offset + buffer_size + element_offset]
-
-        float4* src4 = (float4*)&input[element_offset];
-
-        float4 value;
-        // Check if we have enough elements remaining for a safe float4 load
-        if (element_offset >= 0 && element_offset + ELTS_PER_THREAD <= flags.buffer_size) {
-          value = loadfloat4(src4);
-        } else {
-          // Use safe load for boundary cases or out-of-bounds
-          int remaining_elements = flags.buffer_size - element_offset;
-          if (remaining_elements <= 0) {
-            // Completely out of bounds, return zeros
-            float4 return_value = {0.0f, 0.0f, 0.0f, 0.0f};
-            value = return_value;
-          } else {
-            value = loadfloat4_safe(reinterpret_cast<T_IN const*>(src4), remaining_elements);
-          }
-        }
+  T* scatterBufLocal =
+      reinterpret_cast<T*>(flag.getCurLamportBuf(inputPtrs[rank], MNNVLTwoShotStage::SCATTER));
+  T* scatterBufDest =
+      reinterpret_cast<T*>(flag.getCurLamportBuf(inputPtrs[destRank], MNNVLTwoShotStage::SCATTER));
+  T* broadcastBufW =
+      reinterpret_cast<T*>(flag.getCurLamportBuf(mcastPtr, MNNVLTwoShotStage::BROADCAST));
+  T* broadcastBufR =
+      reinterpret_cast<T*>(flag.getCurLamportBuf(inputPtrs[rank], MNNVLTwoShotStage::BROADCAST));
 
-        if (LAMPORT) {
-          // Assume that the 16B were written atomically, so we only need to check one value
-          T_IN lowest_val = *(T_IN*)&value;
-          valid &= !isNegZero(lowest_val);
-        }
-        *dst4 = value;
-      }
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 900))
+  cudaTriggerProgrammaticLaunchCompletion();
+#endif
+  // Make sure the clear function is called before OOB thread exits
+  if (packedIdx * kELTS_PER_THREAD >= tokenDim) {
+    flag.clearDirtyLamportBuf(inputPtrs[rank], -1);
+    return;
+  }
+
+  // =============================== Scatter ===============================
+
+  // Load vectorized data
+  PackedVec<PackedType, T> val;
+  val.packed = loadPacked<PackedType>(&shardPtr[threadOffset]);
+#pragma unroll
+  for (int i = 0; i < kELTS_PER_THREAD; i++) {
+    if (isNegZero(val.elements[i])) {
+      val.elements[i] = fromFloat<T>(0.F);
     }
   }
 
-  __syncthreads();
+  // Store vectorized data
+  reinterpret_cast<PackedType*>(
+      &scatterBufDest[destTokenOffset * tokenDim * WorldSize + rank * tokenDim])[packedIdx] =
+      val.packed;
 
-  // Perform the initial input reduction
-  if (NUM_INPUTS > 0) {
-    T_IN accum[ELTS_PER_THREAD];
-    float4* accum4 = (float4*)&accum;
+  flag.clearDirtyLamportBuf(inputPtrs[rank], MNNVLTwoShotStage::SCATTER);
```

Comment:
Instead of clearing the buffer here, could we assume we have properly initialized lamport buffers at the start? 
And then, at the end (e.g., right after we call PDL), we can clear the buffers, so that a next kernel using the same workspace can also assume properly initialized lamport buffers.